[compiler-v2] Refactoring of ability instrumentation and checking

wrwg opened the pull request:
This refactors instrumentation of copy and drop instructions, and checking of abilities in general, closing #11223, which lead us to generate unnecessary copies. Copy instrumentation is now moved after reference safety as borrow information is needed to compute it. The existing ability checking and explicit drop processors have been marged into a new `ability_processor` which deals with instrumenting copies and drops as well as checking ability conformance.

The PR also fixes a bug in safety analysis regarding diverging edges, see the new content at the end of file header.

Closes #11223
Closes #11806
Closes #11671
Closes #11924

wrwg commented:
Connect ExitStateAnalysis with last commit. @fEst1ck

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/exit_state_analysis.rs:
```diff
@@ -70,40 +70,25 @@ impl Display for ExitStatus {
-
-#[derive(Clone)]
-pub struct ExitStateAnnotation(BTreeMap<CodeOffset, ExitStateAtCodeOffset>);
+pub struct ExitStateAnnotation(BTreeMap<CodeOffset, ExitState>);
```
Mention that this is the exit state before the instruction at the code offset.

vineethk commented on third_party/move/move-compiler-v2/src/experiments.rs:
```diff
@@ -26,4 +26,7 @@ impl Experiment {
     /// A flag which allows to turn off safety checks, like reference safety.
```
This comment says reference safety is a permanent part of no-safety experiment, but this PR makes reference safety unconditional (not dependent on this flag).

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -239,23 +227,23 @@ impl TestConfig {
+            pipeline.add_processor(Box::new(ReferenceSafetyProcessor {}));
+            pipeline.add_processor(Box::new(ExitStateAnalysisProcessor {}));
+            pipeline.add_processor(Box::new(AbilityProcessor {}));
             pipeline.add_processor(Box::new(AvailCopiesAnalysisProcessor {})); // 0
```
```suggestion
            pipeline.add_processor(Box::new(AvailCopiesAnalysisProcessor {})); // 4
```

brmataptos replied:
@vineethk You know, this mechanism would be less risky if you (1) added a parameter to `add_processor` for the index, and (2) check that you wind up with a contiguous set.

vineethk replied:
@brmataptos Instead of manually adding index and then checking for contiguity, we could just have a boolean param to indicate whether or not to dump based on this processor. I will follow up with a refactor after these PRs land.

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -239,23 +227,23 @@ impl TestConfig {
+            pipeline.add_processor(Box::new(ExitStateAnalysisProcessor {}));
+            pipeline.add_processor(Box::new(AbilityProcessor {}));
             pipeline.add_processor(Box::new(AvailCopiesAnalysisProcessor {})); // 0
             pipeline.add_processor(Box::new(CopyPropagation {})); // 1
```
```suggestion
            pipeline.add_processor(Box::new(CopyPropagation {})); // 5
```

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -239,23 +227,23 @@ impl TestConfig {
-                with_copy_inference: false,
-            }));
+            pipeline.add_processor(Box::new(LiveVarAnalysisProcessor {}));
             pipeline.add_processor(Box::new(DeadStoreElimination {})); // 3
```
```suggestion
            pipeline.add_processor(Box::new(DeadStoreElimination {})); // 7
```

wrwg replied:
Done

vineethk commented on third_party/move/move-compiler-v2/tests/testsuite.rs:
```diff
@@ -239,23 +227,23 @@ impl TestConfig {
                 dump_annotated_targets: true,
                 // Only dump with annotations after these pipeline stages.
-                dump_for_only_some_stages: Some(vec![0, 1, 3]),
+                dump_for_only_some_stages: Some(vec![2, 3, 4, 5]),
```
```suggestion
                dump_for_only_some_stages: Some(vec![4, 5, 7]),
```

wrwg replied:
Done

vineethk commented on third_party/move/move-model/bytecode/src/stackless_bytecode.rs:
```diff
@@ -475,15 +475,13 @@ impl Bytecode {
             Bytecode::Ret(..)
                 | Bytecode::Jump(..)
                 | Bytecode::Abort(..)
+                | Bytecode::Branch(..)
```
It is confusing (and incorrect, IMO) to call a `Branch` with two targets that depends on a condition an "unconditional branch".

brmataptos replied:
It is however true that `Branch` ends a basic block, as there is no fall-through control flow.  Perhaps `is_unconditional_branch` could be renamed to `always_redirects_control_flow` and and `is_conditional_branch` could be renamed to `may_redirect_control_flow`.

wrwg replied:
Renamed

vineethk commented on third_party/move/move-model/bytecode/src/stackless_bytecode.rs:
```diff
@@ -475,15 +475,13 @@ impl Bytecode {
-            self,
-            Bytecode::Branch(..) | Bytecode::Call(_, _, _, _, Some(_))
-        )
+        matches!(self, Bytecode::Call(_, _, _, _, Some(_)))
```
Same comment as above: why is `Branch` not a conditional branch?

wrwg replied:
Done

vineethk commented on third_party/move/move-model/src/ty.rs:
```diff
@@ -2209,14 +2223,15 @@ where
+                    );
                 }
                 if is_phantom_position {
                     // phantom type parameters don't participte in ability derivations
```
```suggestion
                    // phantom type parameters don't participate in ability derivations
```

wrwg replied:
Done

brmataptos commented on third_party/move/move-compiler-v2/src/experiments.rs:
```diff
@@ -4,7 +4,7 @@
 /// Container for experiments in the compiler. Those can be activated
 /// via the `--experiment=<name>` flag. One can also use the env var
-/// `MOVE_COMPILER_EXP=<comma list>` to activate those flags.
+/// `MVC_EXP=<comma list>` to activate those flags.
```
If you change compiler controls, please update the best doc we have currently: https://www.notion.so/aptoslabs/Move-Compiler-Debugging-Tips-322ae742b44846da840eec0f4e195a0b#023e370ac7264eebb4215d42d3e545b7

which should be cleaned up and put into a README, I guess.  (I filed an issue to track that https://github.com/aptos-labs/aptos-core/issues/12163).

wrwg replied:
Really should be in markdown than this would be smoother. In any case, the document is in `experiment.rs` and has been updated, the other doc should just point to it I think.

wrwg replied:
Your irony doesn't make a lot of sense here. The source of truth is documentation on code or markdown files. Experiments are a standard concept of our compiler and they are documented in this file. There is not need to go of thousands of lines of code since this here has only a few dozens here.

brmataptos reviewed (commented):
Haven't looked at ability_processor yet.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:
```diff
@@ -0,0 +1,656 @@
+        // Only temps which are used after or borrowed need a copy
+        let temp_needs_copy =
+            |temp| live_var.after.contains_key(temp) || lifetime.before.is_borrowed(*temp);
+        // References always need to be dropped to satisfy bytecode verifier borrow analysis, other values
```
Fair enough.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:
```diff
@@ -0,0 +1,656 @@
+            },
+            Call(_, _, _, srcs, ..) => {
+                // For arguments, we also need to check the case that a src is not used after this program point
+                // but again in the argument list. Also, in difference to assign, we only need to copy the
```
I thought this was referring to the code: `temp_needs_copy(src) || srcs[i + 1..].contains(src)`. I suggested "or" because of this. Re-reading this, I think both would be fine. Ok to leave this as is.

wrwg replied:
its the 'not used after but...' where you scanned 'used after or ..'. I changed it to avoid the confusion.

vineethk commented on third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:
```diff
@@ -0,0 +1,656 @@
+    fn check_and_emit_bytecode(&mut self, _code_offset: CodeOffset, bc: Bytecode) {
+        use Bytecode::*;
+        #[allow(clippy::single_match)]
+        match &bc {
```
Gotcha. I think it would make sense to globally disable if we are frequently adding such exceptions.

I have not tested this, but from reading the lint rule (https://rust-lang.github.io/rust-clippy/master/#/single_match), it seems like if you add any comment in the other match arm (_), then the lint wont complain. That might be another way to go.

vineethk reviewed (approved):
LGTM. Note: I have not carefully reviewed all the test outputs, especially given changes here combined with the other PR this is diffed against.

wrwg reviewed (commented):
This PR has been rebased and aligned with the downstream changes. PTAL

fEst1ck commented:
Not sure if I missed something. But `ability-check/operators.move` is empty? I tried the original `operators.move` and `fields.move` but https://github.com/aptos-labs/aptos-core/issues/11924 is still not fixed?

brmataptos commented on third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:
```diff
@@ -0,0 +1,653 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! The ability processor checks conformance to Move's ability system as well as instruments
```
Using "instrument" here is confusing.  To "instrument" is to "equip (something) with measuring instruments".  It especially gets confusing below since "instrumentation" (in software, anyway) can mean either the thing adding instruments, or the instruments themselves.

If you love "instrument", then on line 7, you *must* change it to "instrumentation process" or "instrumenter" or something to distinguish the processor from the things inserted.

But I think you should change it to "<something> inference".  Here we are inferring the Copy, Move, and Drop operations which are implicit in the source code.  I wish there were a single name that encapsulated those 3 attributes, but in discussions I've found of Rust, there is some argument that:
- Every assignment is a Move
- An object which has Copy trait just leaves a copy behind

Extending that to Move, 
- By default, every assignment is either a Move or a Copy depending on whether type has Copy
- This analysis simply infers needed Drop operations:
   - a Copy assignment followed by a Drop of the source simplifies to a Move assignment
   - in other situations, we need an explicit Drop operation.

But this doesn't quite jive with Move, which has explicit "move" and "copy" assignments.  The docs even refer to "a move or a copy would be inserted by the compiler".  But calling this "Drop inference"  or "AssignmentDrop inference" would be best for us implementers.

See https://aptos.dev/move/book/variables/#move-and-copy

wrwg replied:
Ok. I've seen instrumentation in this context many times, not just for measuring, but I changed it to "transform" since it seems important to you.

brmataptos commented on third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:
```diff
@@ -0,0 +1,656 @@
+                    | MoveTo(mod_id, struct_id, insts) => {
+                        self.check_key_for_struct(*id, *mod_id, *struct_id, insts)
+                    },
+                    Drop => self.check_drop(*id, srcs[0], || {
```
It's not in the parser.  You can always write a function to do it.

vineethk replied:
Who does the "you" refer to in the final sentence? Is it the developer? If so, I do not quite understand - if you drop something in a function, and call that function, that does not result in a `Drop` bytecode here ...

Also, clarifying that there is no action item for the PR here, I just wanted to understand @brmataptos' comment.

brmataptos replied:
"you" = "the developer".  Yes, there is apparently no way to create a `Drop` bytecode from Move source language.  The *effect* of a `Drop` bytecode can be achieved by a function call that takes a ref parameter and implicitly drops it, though that won't work for an `inline` function.

brmataptos reviewed (commented):
Looks good.  A few naming and comment nits.

brmataptos commented on third_party/move/move-compiler-v2/tests/live-var/mut_ref.exp:
```diff
@@ -162,14 +162,6 @@ fun m::some2($t0: &mut m::R, $t1: &mut m::R) {
-error: implicit copy of mutable reference in local `x` which is used later in argument list
-   ┌─ tests/live-var/mut_ref.move:41:9
-   │
-41 │         some2(x, x); // expected error because multiple use
```
We seem to be losing this error in multiple test cases.

wrwg replied:
Livevar doesn't longer check this, so its legit.

wrwg replied:
It is now in reference_safety/multiple_use_invalid, and also a bit more extensively tested, specifically involving reference aliasing.

brmataptos replied:
Ok.

brmataptos reviewed (commented):
Ok, I checked all the tests except the copy-propagation tests which are seemingly all problematic.

We have a few problematic missing errors, notably `foo(t1, t2)` where both params are `&mut s`.

wrwg commented on third_party/move/move-compiler-v2/tests/ability-check/bad_inst.exp:
```diff
@@ -0,0 +1,13 @@
+
+Diagnostics:
+error: type `test::T` is missing ability `drop`
```
This will be fixed with #12036

wrwg commented on third_party/move/move-compiler-v2/tests/ability-check/inferred_copy.exp:
```diff
@@ -0,0 +1,9 @@
+
+Diagnostics:
+error: cannot move local `r` since it is still used
```
Later is not always correct, that is why I changed this. It can now be also used because it has been previously borrowed and still is.

wrwg replied:
Done

brmataptos replied:
Nice.

wrwg commented on third_party/move/move-compiler-v2/tests/ability-check/invalid_struct_def.exp:
```diff
@@ -1,55 +1,55 @@
 
-error: Invalid instantiation
-  ┌─ tests/ability-checker/invalid_struct_def.move:6:18
+error: missing ability `key`
```
With #12036 this should be fixed. The problem is really that the location is wrong which this bug addresses.

wrwg replied:
This is not a test failure, please look closer. This replaces the error message 'invalid instantiation' with 'missing ability `key`', which is certainly better than before. The problem is that this should have pointed to `u8` not to `x`, as I tried to say above.

And this is covered by #12036. Once this is implemented, we will migrate those tests into the frontend and then re-inspect them.

wrwg commented on third_party/move/move-compiler-v2/tests/ability-check/operators.move:
Merge issue, fixed

wrwg commented on third_party/move/move-compiler-v2/tests/reference-safety/v1-tests/borrow_global.exp:
```diff
@@ -0,0 +1,14 @@
+
+Diagnostics:
+error: cannot immutably borrow since mutable references exist
```
https://github.com/aptos-labs/aptos-core/issues/12214

wrwg replied:
Attributed the bug wrongly. The errors are different between v1 and v2, this is legit in the logic of v2.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:
```diff
@@ -710,27 +733,56 @@ impl LifetimeState {
     fn roots(&self, label: &LifetimeLabel) -> BTreeSet<LifetimeLabel> {
         let mut roots = BTreeSet::new();
         let mut todo = self.node(label).parents.iter().cloned().collect::<Vec<_>>();
```
Don't micro-optimize my code please, for a couple of nano seconds. I prefer the more readable version.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:
```diff
@@ -710,27 +733,56 @@ impl LifetimeState {
+    /// Returns the transitive children of this node.
+    fn transitive_children(&self, label: &LifetimeLabel) -> BTreeSet<LifetimeLabel> {
+        // Helper function to collect the target nodes of the children.
+        let get_children =
```
Because the lambda is more concise since of inference, its rather idiomatic.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:
```diff
@@ -710,27 +733,56 @@ impl LifetimeState {
+            result.insert(*label);
+        } else {
+            while let Some(l) = todo.pop() {
+                if !result.insert(l) {
```
I like it how it is. Please do not insert your personal taste in code reviews.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:
```diff
@@ -1258,7 +1326,16 @@ impl<'env, 'state> LifetimeAnalysisStep<'env, 'state> {
                 AssignKind::Move => self.state.move_ref(dest, src),
                 AssignKind::Copy => self.state.copy_ref(dest, src),
-                _ => panic!("assign kind expected to be Move or Copy"),
+                AssignKind::Inferred => {
```
is_borrowed is a public function on the annotation not the analysis state.

wrwg commented on third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:
```diff
@@ -1492,11 +1568,17 @@ impl<'env> TransferFunctions for LifeTimeAnalysis<'env> {
+            // Only handle operations which can take references
             Call(_, _, oper, srcs, ..) => match oper {
-                Operation::ReadRef | Operation::WriteRef | Operation::Function(..) => {
+                Operation::ReadRef
```
I know this and use it all the time, decided to do it like this.

Please do not micromanage my code.

brmataptos replied:
Don't worry, no one is micromanaging your code.  I'm offering a potentially helpful fact about Rust.  If you already know it you can just say so.

wrwg commented on third_party/move/move-compiler-v2/tests/ability-check/globals.move:
Merge issue, done

wrwg commented on third_party/move/move-compiler-v2/tests/reference-safety/v1-tests/call_mutual_borrows_invalid.exp:
```diff
@@ -47,37 +43,24 @@ error: cannot mutably borrow since immutable references exist
-13 │     fun t0(s1: &mut S, _s2: &mut S) {
-   │            -- previous mutable local borrow
-   ·
-23 │         mut_mut(s1, s1);
```
https://github.com/aptos-labs/aptos-core/issues/12214

wrwg commented:
Comments addressed, PTAL

wrwg commented:
> Not sure if I missed something. But `ability-check/operators.move` is empty? I tried the original `operators.move` and `fields.move` but #11924 is still not fixed?

Result of a merge conflicts, brought them back

wrwg commented:
Comments addressed, PTAL

brmataptos commented on third_party/move/move-compiler-v2/tests/ability-check/ability_violation.exp:
```diff
@@ -0,0 +1,75 @@
+   ┌─ tests/ability-check/ability_violation.move:4:3
+   │
+ 4 │         move_from<T>(addr);
+   │         ^^^^^^^^^^^^^^^^^^ implicitly dropped here since not longer used
```
Sorry to keep nitpicking, but "since not longer used" is painful.  "since it is no longer used" (best) or (at least) "since no longer used", please.

wrwg replied:
Done

brmataptos commented on third_party/move/move-compiler-v2/tests/ability-check/operators.move:
Still seems missing.

wrwg replied:
There are no errors so the exp isn't produced.

brmataptos reviewed (commented):
Please double-check that you're not missing an .exp file or two.

fEst1ck commented on third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:
```diff
@@ -0,0 +1,653 @@
+        use Bytecode::*;
+        #[allow(clippy::single_match)] // For handling of future cases
+        match &bc {
+            Call(id, _, op, srcs, _) => {
```
Where did we check drop abilities for eq/neq operations? Or we don't check it here?

wrwg replied:
There is nothing special about eq/neq, they behave like every other function. See also `CopyDropAnalysis::execute`

brmataptos reviewed (approved):
Nothing crucial.  It would be good if we can clean up the error messages a tad, but up to you.

wrwg commented:
I added a fix for #11924, and addressed the error message naming.  For clarity, there are now independent tests covering this bug: `alive_since_borrowed.move` and `operator_drop`.  Thank you for pushing the issue with equalities @fEst1ck : this is actually interesting since this function behaves logical like every other (consumes arguments) but has this additional error so the user does not shoot in there foot. This is bad devx and the function should instead automatically take references for the arguments. We can change the language here perhaps later.
